src/store/Database.test.ts

```diff
@@ -0,0 +1,679 @@
+import { IDBPDatabase, deleteDB } from "idb";
+
+import { expectPromise } from "../__tests__/helpers/expect";
+
+import { Database, TransactionMode } from "./Database";
+import { NamedSchema } from "./types";
+
+const testDBName = "databaseTestDB";
+const testStoreName = "testStore";
+const anotherTestStoreName = "anotherTestStore";
+
+type TestSchema = NamedSchema<
+  typeof testDBName,
+  {
+    [testStoreName]: {
+      key: string;
+      value: {
+        a: string;
+        b: number;
+      };
+    };
+
+    [anotherTestStoreName]: {
+      key: string;
+      value: number;
+    };
+  }
+>;
+
+let initialDB: Database<TestSchema>;
+let db: Database<TestSchema>;
+
+const createDBWithStore = async (
+  version = 1
+): Promise<Database<TestSchema>> => {
+  return (db = await Database.open<TestSchema>(testDBName, version, {
+    upgrade(upgrade) {
+      upgrade.createStore(testStoreName);
+      upgrade.createStore(anotherTestStoreName);
+    }
+  }));
+};
+
+afterEach(async () => {
+  if (initialDB) {
+    initialDB.close();
+  }
+
+  if (db) {
+    db.close();
+  }
+
+  await deleteDB(testDBName);
+});
+
+describe(".open()", () => {
+  it("opens a connection to an IndexedDB database with the provided name and schema version", async () => {
+    const version = 5;
+    const spy = jest.spyOn(indexedDB, "open");
+
+    db = await Database.open<TestSchema>(testDBName, version);
+
+    expect(spy).toHaveBeenCalledTimes(1);
+    expect(spy).toHaveBeenCalledWith(testDBName, version);
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const internalDB = (db as any).db as IDBPDatabase<TestSchema>;
+
+    expect(internalDB).toBeInstanceOf(IDBDatabase);
+    expect(internalDB.name).toEqual(testDBName);
+    expect(internalDB.version).toEqual(version);
+  });
+
+  it("throws when attempting to reopen a database with an older schema version", async () => {
+    const initialVersion = 7;
+    const currentVersion = 4;
+
+    initialDB = await Database.open<TestSchema>(testDBName, initialVersion);
+
+    initialDB.close();
+
+    await expectPromise(async () => {
+      db = await Database.open<TestSchema>(testDBName, currentVersion);
+    }).rejects.toThrow();
+  });
```

Comment:
Do we need to display an error here?

Reply:
Yes, we should be testing that. Thanks.

Reply:
Ah. The reason for this is that the error itself doesn't actually have a message. We might want to change that, but I don't think it's a priority.